### PR TITLE
4 - edit user form

### PR DESCRIPTION
**Pull request by rfontanarosa**
part of #1388

https://user-images.githubusercontent.com/10047945/181475312-959d10f0-6a9d-43a6-9fad-2bb470eb5ea2.mov

### CODE REVIEW

File: client/components/EditUserForm/Buttons/Buttons.tsx [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  const i18n = useTranslation()
+  const countryIso = useCountryIso()
+  const userInfo = useUser()
+  const showDeactivate =
+    Users.isAdministrator(userInfo) || (Users.isNationalCorrespondent(userInfo, countryIso) && user.id !== userInfo.id)
+
+  return (
+    <>
+      {showDeactivate && (
+        <div className="edit-user__form-item-buttons">
```
Comment (minotogna) [inline client/components/EditUserForm/Buttons/Buttons.tsx:28]: move sass code related to this component to local file

File: client/components/EditUserForm/EditUserForm.tsx [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+  const { toaster } = useToaster()
+  const countryIso = useCountryIso()
+  const assessment = useAssessment()
+  const cycle = useCycle()
+
+  const [profilePicture, setProfilePicture] = useState<File>(null)
+
+  if (!user) return null
+
+  const onSave = () => {
```
Comment (minotogna) [inline client/components/EditUserForm/EditUserForm.tsx:33]: use preferably useCallback

File: client/components/EditUserForm/EditUserForm.tsx [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        dispatch(
+          UserManagementActions.getUsers({ countryIso, assessmentName: assessment.props.name, cycleName: cycle.name })
+        )
+        dispatch(UserManagementActions.setUserToEdit(null))
+        toaster.success(i18n.t('userManagement.userModified', { email: user.email }))
+      })
+    }
+  }
+
+  const onChange = (value: string, key: string) => {
```
Comment (minotogna) [inline client/components/EditUserForm/EditUserForm.tsx:45]: use preferably useCallback

File: client/components/EditUserForm/EditUserForm.tsx [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+  return (
+    <div className="edit-user__form-container">
+      <ProfilePicture userId={user.id} onChange={(profilePicture: File) => setProfilePicture(profilePicture)} />
+      <TextInputFields user={user} onChange={onChange} />
+      {/* <CountryRoles onChange={onChange} user={user} /> */}
+
+      <Buttons
+        user={user}
+        userActive={user.status === UserStatus.active}
+        onDeactivate={() =>
```
Comment (minotogna) [inline client/components/EditUserForm/EditUserForm.tsx:63]: use preferably useCallback  and extract callbacks to variables

File: client/pages/AssessmentHome/FraHome/Collaborators/Collaborators.tsx [hunk 5]
```diff
@@ -1,20 +1,41 @@
...
-  return (
+  useEffect(() => {
+    dispatch(UserManagementActions.setUserToEdit(null))
+  }, [])
+
+  useEffect(() => {
+    dispatch(
+      UserManagementActions.getUsers({ countryIso, assessmentName: assessment.props.name, cycleName: cycle.name })
+    )
+  }, [countryIso])
```
Comment (minotogna) [inline client/pages/AssessmentHome/FraHome/Collaborators/Collaborators.tsx:34]: wasn't this somewhere else ?

File: common/api/endpoint/ApiEndPoint.ts [hunk 6]
```diff
@@ -102,6 +102,7 @@ export const ApiEndPoint = {
   },
 
   User: {
+    one: () => apiPath('user'),
```
Comment (minotogna) [inline common/api/endpoint/ApiEndPoint.ts:105]: this is in reality 'many'.
also, rename first users api path to 'users'

File: meta/user/index.ts [hunk 7]
```diff
@@ -1,19 +1,18 @@
...
+  UserRole,
   Viewer,
-  CollaboratorProps,
-  Collaborator,
 } from './userRole'
-
+export { RoleName } from './userRole'
 export { Users } from './users'
-export { Authorizer } from './authorizer'
+export { UserUtils } from './userUtils'
```
Comment (minotogna) [inline meta/user/index.ts:18]: Users is already UserUtils. remove UserUtils

File: server/api/user/updateUser.ts [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+export const updateUser = async (req: Request, res: Response) => {
+  try {
+    const profilePicture = req.file
+    const userToUpdate = JSON.parse(req.body.user)
+
+    if (profilePicture) {
+      await UserController.updateProfilePicture({ user: userToUpdate, file: profilePicture })
+    }
+
+    const updatedUser = await UserController.update({ user: userToUpdate })
```
Comment (minotogna) [inline server/api/user/updateUser.ts:15]: pass profilePicture in UserRepository.update . one method is enough

File: server/api/user/index.ts [hunk 9]
```diff
@@ -20,5 +23,6 @@ export const UserApi = {
     express.post(ApiEndPoint.User.invite(), AuthMiddleware.requireInviteUser, invite)
     express.delete(ApiEndPoint.User.removeInvitation(), AuthMiddleware.requireInviteUser, removeInvitation)
     express.get(ApiEndPoint.User.sendInvitationEmail(), AuthMiddleware.requireInviteUser, sendInvitationEmail)
+    express.put(ApiEndPoint.User.one(), multer().single('profilePicture'), updateUser)
```
Comment (minotogna) [inline server/api/user/index.ts:26]: add middleware requireEditUser

**Review by minotogna — CHANGES_REQUESTED** [review]

File: client/pages/AssessmentHome/FraHome/Collaborators/Collaborators.tsx (hunk 5 above)
Comment (rfontanarosa) [reply client/pages/AssessmentHome/FraHome/Collaborators/Collaborators.tsx:34]: yes, it was into FraHome. I removed it from FraHome and keep it here in Collaborators

File: client/pages/AssessmentHome/FraHome/FraHome.tsx [hunk 10]
```diff
@@ -20,24 +16,12 @@ import SelectedCountries from './SelectedCountries'
...
   const displayTabs = sections.length > 1 && Areas.isISOCountry(countryIso)
 
-  useEffect(() => {
-    if (user) {
-      dispatch(
-        UserManagementActions.getUsers({ countryIso, assessmentName: assessment.props.name, cycleName: cycle.name })
-      )
-    }
-  }, [countryIso, cycle, assessment, user])
-
```
Comment (minotogna) [inline client/pages/AssessmentHome/FraHome/FraHome.tsx:40]: check if this breaks Home/message board

Comment (rfontanarosa) [reply client/pages/AssessmentHome/FraHome/FraHome.tsx:40]: :( you're right

File: client/components/EditUserForm/EditUserForm.tsx [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+          ...user,
+          [key]: value,
+        })
+      )
+    },
+    [dispatch, user]
+  )
+
+  const deactivateUser = useCallback(() => {
+    dispatch(
```
Comment (minotogna) [inline client/components/EditUserForm/EditUserForm.tsx:53]: Does deactivate set in db active to false ?

**Review by minotogna — APPROVED** [review]

File: client/components/EditUserForm/EditUserForm.tsx (hunk 11 above)
Comment (rfontanarosa) [reply client/components/EditUserForm/EditUserForm.tsx:53]: yes, it does

**Pull request merged**